File: fbhc/2021/round1/A2.js
const sigma = (a, b, f) => {
  sum = 0;
  for (let i = a; i <= b; i++) sum += f(i);
  return sum;
};

// const t = [0, 0, 2, 1, 1, 2];
// const a = [0, 1, 2, 1, 1];
// const k = 4;

// const ans = sigma(1, k, i => (a[i] + t[i+1]) * sigma(1, i-1, j => (i - j) * (t[j] + a[j])));

// for (let i = 4; i >= 1; i--) {
//   const c = sigma(1, i-1, j => (i - j) * (t[j] + a[j]));
//   console.log(`(${(a[i] + t[i+1])}) * ${c}`);
//   for (let j = 1; j < i; j++) {
//     console.log(`\t${i-j} * ${t[j]+a[j]}`);
//   }
// }

// console.log(ans);

const t = [0, 0, 1, 1];
const a = [0, 1, 1];
const k = 2;

const ans = 2 + t[1] * 


sigma(1, k, i => (a[i] + t[i+1]) * sigma(1, i-1, j => (i - j) * (t[j] + a[j]))) +
  sigma(1, k, i => (a[i] + t[i+1]) * (
    sigma(1, k, j => (k + i - j) * (t[j] + a[j])) +
      t[k+1] * (i-1) +
      sigma(1, i-1, j => (i-j) * (t[j] + a[j]))));

console.log(ans);

const F = sigma(1, k-1, j => (k - j) * (t[j] + a[j]));
const S = sigma(1, k, j => j * (t[j] + a[j]));
console.log('F', F);
console.log('S', S);

const X = sigma(1, k, i => (t[i+1] + a[i]) * (i-1));

File: fbhc/2021/round1/A2.cpp
#include <bits/stdc++.h>
using namespace std;

#define MOD 1000000007LL

int n;
char s[800800];

template<typename T> class BIT {
public:
  BIT(int n): v(vector<T>(n+1)), n(n) {}
  T sum(int k) {
    T s = 0;
    for (int i = k; i > 0; i -= (i & -i)) {
      s += this->v[i];
    }
    return s;
  }
  T sum(int l, int r) {
    return this->sum(r) - this->sum(l-1);
  }
  void update(int k, T v) {
    for (int i = k; i <= this->n; i += (i & -i)) {
      this->v[i] += v;
    }
  }
private:
  vector<T> v;
  int n;
};

void solve() {
  scanf("%d", &n);
  scanf("%s", s);

  BIT<long long> bit(n);
  long long ans = 0, fixed = 0, temp = 0;
  int k = 0;
  char last = 'F';
  for (int i = 0; i < n; ++i) {
    if (s[i] == 'F') {
      temp++;
    } else {
      if (last != s[i]) {
        fixed += bit.sum(k);
        fixed %= MOD;
        k++;
      }
      bit.update(k, temp+1);
      temp = 0;
      last = s[i];
    }
    ans += fixed;
    ans %= MOD;
    // printf("[%d] = %d:", i, fixed);
    // for (int i = 1; i <= k; ++i) {
    //   printf(" %d", bit.sum(i, i));
    // }
    // printf(" ? %d\n", temp);
  }

  printf("%lld\n", ans);
}

int main() {
  int ntc;
  scanf("%d", &ntc);
  for (int itc = 1; itc <= ntc; ++itc) {
    printf("Case #%d: ", itc);
    solve();
  }
}